### PR TITLE
[C# Port] Adding new Proxy Update Queue Function.

### PR DESCRIPTION
**Pull request by nharper285**
## Summary of the Pull Request

_What is this about?_
Porting QueueProxyUpdate to C# for refactor. 

## PR Checklist
* [x] Applies to work item: #1748 
* [x] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_
New: 
* QueueProxyUpdate. New function file with the refactored function changes.
Changed: 
* models.cs - Added Proxy, ProxyForward, and ProxyHeartbeat models

## Validation Steps Performed

_How does someone test & validate?_
I deployed the function locally and connected it to the proxy queue in an active instance. The function properly processed the entries on the queue.

### CODE REVIEW

File: src/ApiService/ApiService/OneFuzzTypes/Model.cs [hunk 1]
```diff
@@ -87,6 +94,26 @@ bool DebugKeepNode
 ) : EntityBase();
 
 
+public partial record ProxyForward
+(
+	string Region,
+	int SrcPort,
+	int DstPort,
+	int DstIp
+) : EntityBase();
```
Comment (chkeita) [inline src/ApiService/ApiService/OneFuzzTypes/Model.cs:103]: you don't need to derive form EntityBase if this is not meant to be stored in the db

File: src/ApiService/ApiService/QueueProxyUpdate.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+{
+    private readonly ILogger _logger;
+
+    public QueueProxyUpdate(ILoggerFactory loggerFactory)
+    {
+        _logger = loggerFactory.CreateLogger<QueueProxyUpdate>();
+    }
+
+    [Function("QueueProxyUpdate")]
+    public void Run([QueueTrigger("myqueue-items", Connection = "AzureWebJobsStorage")] string msg)
```
Comment (chkeita) [inline src/ApiService/ApiService/QueueProxyUpdate.cs:19]: this needs to be update to match the config in host.json

File: src/ApiService/ApiService/OneFuzzTypes/Model.cs (hunk 1 above)
Comment (nharper285) [reply src/ApiService/ApiService/OneFuzzTypes/Model.cs:103]: Proxy Forward is currently stored in the Azure tables.

File: src/ApiService/ApiService/QueueProxyUpdate.cs (hunk 2 above)
Comment (nharper285) [reply src/ApiService/ApiService/QueueProxyUpdate.cs:19]: Yes was just doing this for testing!

File: src/ApiService/ApiService/OneFuzzTypes/Model.cs [hunk 3]
```diff
@@ -88,6 +102,40 @@ bool DebugKeepNode
...
+	[PartitionKey] string Region,
+	[RowKey] int DstPort,
+    int SrcPort,
+	string DstIp
+) : EntityBase();
+
+public partial record ProxyConfig 
+(	
+	string Url,
+	string Notifcation,
```
Comment (tevoinea) [inline src/ApiService/ApiService/OneFuzzTypes/Model.cs:116]: Notif**i**cation

File: src/ApiService/ApiService/OneFuzzTypes/Model.cs [hunk 4]
```diff
@@ -88,6 +102,40 @@ bool DebugKeepNode
...
+(
+	[PartitionKey] string Region,
+	[RowKey] int DstPort,
+    int SrcPort,
+	string DstIp
+) : EntityBase();
+
+public partial record ProxyConfig 
+(	
+	string Url,
```
Comment (tevoinea) [inline src/ApiService/ApiService/OneFuzzTypes/Model.cs:115]: We can make this a URI? https://docs.microsoft.com/en-us/dotnet/api/system.uri?view=net-6.0

File: src/ApiService/ApiService/.gitignore [hunk 5]
```diff
@@ -0,0 +1,264 @@
+## Ignore Visual Studio temporary files, build results, and
+## files generated by popular Visual Studio add-ons.
```
Comment (chkeita) [inline src/ApiService/ApiService/.gitignore:2]: This file is not needed . we already have the same under src\ApiService

File: src/ApiService/ApiService/OneFuzzTypes/Model.cs [hunk 6]
```diff
@@ -1,5 +1,7 @@
 using Microsoft.OneFuzz.Service.OneFuzzLib.Orm;
 using System;
+using System.Collections.Generic;
+using System.Net;
```
Comment (chkeita) [inline src/ApiService/ApiService/OneFuzzTypes/Model.cs:4]: make sure to run `dotnet format` in src\ApiService to remove those formatting issues

**Review by chkeita — APPROVED** [review]

File: src/ApiService/Tests/OrmTest.cs [hunk 7]
```diff
@@ -109,6 +111,7 @@ public void TestFromtableEntity()
             Assert.Equal(tableEntity.GetDouble("the_float"), entity1.TheFloat);
             Assert.Equal(TestEnum.TheTwo, entity1.TheEnum);
             Assert.Equal(tableEntity.GetString("a__special__name"), entity1.Renamed);
+            Assert.Equal(tableEntity.GetString("test_null"), null);
```
Comment (chkeita) [inline src/ApiService/Tests/OrmTest.cs:114]: replace with 
``` suggestion 
Assert.Equal(null, tableEntity.GetString("test_null"));
Assert.Equal(tableEntity.GetString("test_null"), entity1.TestNull);
```

File: src/ApiService/ApiService/QueueProxyHeartbeat.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        _proxy = proxy;
+    }
+
+    [Function("QueueProxyHearbeat")]
+    public async Task Run([QueueTrigger("myqueue-items", Connection = "AzureWebJobsStorage")] string msg)
+    {
+        _logger.LogInformation($"heartbeat: {msg}");
+
+        var hb = JsonSerializer.Deserialize<ProxyHeartbeat>(msg, EntityConverter.GetJsonSerializerOptions()).EnsureNotNull($"wrong data {msg}"); ;
+        var newHb = hb with { TimeStamp = DateTimeOffset.UtcNow };
```
Comment (tevoinea) [inline src/ApiService/ApiService/QueueProxyHeartbeat.cs:28]: I didn't know about `with`, neat!

File: src/ApiService/ApiService/onefuzzlib/orm/EntityConverter.cs [hunk 9]
```diff
@@ -245,6 +245,10 @@ private EntityInfo GetEntityInfo<T>()
                     else
                     {
                         var value = entity.GetString(fieldName);
+                        if (value == null)
+                        {
+                            return null;
```
Comment (tevoinea) [inline src/ApiService/ApiService/onefuzzlib/orm/EntityConverter.cs:250]: Shouldn't this also change the type of this function to `T?`?

**Review by tevoinea — APPROVED** [review]

File: src/ApiService/ApiService/onefuzzlib/orm/EntityConverter.cs (hunk 9 above)
Comment (nharper285) [reply src/ApiService/ApiService/onefuzzlib/orm/EntityConverter.cs:250]: ToRecord already has a type `T` return type yeah?

Comment (chkeita) [reply src/ApiService/ApiService/onefuzzlib/orm/EntityConverter.cs:250]: We getting the value of the parameter of record constructor, not record itself.

**Pull request merged**